document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresClient.java

```diff
@@ -0,0 +1,112 @@
+package org.hypertrace.core.documentstore.postgres;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class PostgresClient {
+
+  private static final Logger log = LoggerFactory.getLogger(PostgresClient.class);
+  private static final int VALIDATION_QUERY_TIMEOUT_SECONDS = 5;
+
+  private final String url;
+  private final String user;
+  private final String password;
+  private final int maxConnectionAttempts;
+  private final Duration connectionRetryBackoff;
+
+  private int count = 0;
+  private Connection connection;
+
+  public PostgresClient(
+      String url,
+      String user,
+      String password,
+      int maxConnectionAttempts,
+      Duration connectionRetryBackoff) {
+    this.url = url;
+    this.user = user;
+    this.password = password;
+    this.maxConnectionAttempts = maxConnectionAttempts;
+    this.connectionRetryBackoff = connectionRetryBackoff;
+  }
+
+  public synchronized Connection getConnection() {
+    try {
+      if (connection == null) {
+        newConnection();
+      } else if (!isConnectionValid(connection)) {
+        log.info("The database connection is invalid. Reconnecting...");
+        close();
+        newConnection();
+      }
+    } catch (SQLException sqle) {
+      throw new RuntimeException(sqle);
+    }
+    return connection;
+  }
+
+  private synchronized boolean isConnectionValid(Connection connection) {
+    try {
+      if (connection.getMetaData().getJDBCMajorVersion() >= 4) {
+        return connection.isValid(VALIDATION_QUERY_TIMEOUT_SECONDS);
+      } else {
+        try (PreparedStatement preparedStatement = connection.prepareStatement("SELECT 1");
+            ResultSet resultSet = preparedStatement.executeQuery()) {
+          return true;
+        }
+      }
+    } catch (SQLException sqle) {
+      log.debug("Unable to check if the underlying connection is valid", sqle);
+      return false;
+    }
+  }
+
+  private synchronized void newConnection() throws SQLException {
+    int attempts = 0;
+    while (attempts < maxConnectionAttempts) {
+      try {
+        ++count;
```

Comment:
I think, If, we want to know how many different new connections are used throughout, we should increment the count in `getConnection` after line no: `47`. 

So, between old valid connection = 2, and newValidConenction=3, there were X attempts were made. This information is logged. This will also help in debugging actually which connection was closed, it's kind of id of that connection.

Reply:
FIxed